src/utils/Petsc.hpp

```diff
@@ -62,27 +62,40 @@ class Vector
   /// Creates a new vector on the given MPI communicator.
   explicit Vector(std::string name = "");
 
+  /** Copy construction from another vector
+   * Duplicates the vector and copies the name
+   */
+  Vector(const Vector& other);
+
+  /** Copy assignement
+   * Destroys the current vector and takes ownership of the other.
+   */
+  Vector& operator=(Vector other);
+
+  /** Move construction
+   * Takes ownership of the other vector.
+   */
+  Vector(Vector&& other);
+
   /// Use Vec v as vector.
   Vector(Vec &v, std::string name = "");
 
-  /// Duplicates type, row layout etc. (not values) of v.
-  Vector(Vector &v, std::string name = "");  
-
   /// Constructs a vector with the same number of rows (default) or columns.
   Vector(Mat &m, std::string name = "", LEFTRIGHT type = LEFT);
 
   /// Constructs a vector with the same number of rows (default) or columns.
   Vector(Matrix &m, std::string name = "", LEFTRIGHT type = LEFT);
 
-  /// Delete copy and assignment constructor
-  /** Copying and assignement of this class would involve copying the pointer to
-      the PETSc object and finallly cause double destruction of it.
+  /** Creates an uninitialized vector of identical shape.
+   * Duplicates type, row layout etc. (not values) of v.
    */
-  Vector(const Vector&) = delete;
-  Vector& operator=(const Vector&) = delete;
+  Vector makeSimilar(const std::string& name = "") const;
```

Comment:
This is not really uniform with the previous two methods. I would suggest:
```suggestion
  static Vector makeSimilar(const Vector& templ, const std::string& name = "");
  static Vector makeSimilar(const Matrix& templ, const std::string& name = "", LEFTRIGHT type = LEFT);
  static Vector makeSimilar(const Mat& templ, const std::string& name = "", LEFTRIGHT type = LEFT);
```
Another option would be to move
```
Vector(Matrix &m, std::string name = "", LEFTRIGHT type = LEFT);
```
to `Matrix` and require `Vector(Matrix(Mat))` for the other function.

Reply:
These methods have different concerns:
* The `Vector` method allocates an identically shaped `Vector`
* The `Matrix` methods allocate a compatible `Vector`

The methods require an existing `Vector` and an existing `Matrix`/`Mat` to work with. Thus I would turn them into member functions of the respective dependency.

What do you PETSc guys think looks better/more natural?
```cpp
Vector v;
auto u = v.makeSimilar("u");
auto u = Vector::makeSimilar(v, "u");

Matrix a;
auto w = Vector::makeSimilar(a, Left);
auto w = a.makeCompatible(Left);
auto w = Matrix::makeCompatible(a, Left);
auto w = Vector::makeCompatible(a, Left);
```

Reply:
> These methods have different concerns:

They both create a completely new `Vector`. 

Why not go with simple factory functions and distinguish by overloads only?
```suggestion

  static Vector Vector::instance(const std::string& name = "");
  static Vector Vector::instance(const Vector& templ, const std::string& name = "");
  static Vector Vector::instance(const Matrix& templ, const std::string& name = "", LEFTRIGHT type = LEFT);
  static Vector Vector::instance(const Mat& templ, const std::string& name = "", LEFTRIGHT type = LEFT);
```
When creating a `Vector` from another `Vector` one would assume they should be similar. When constructing from a `Matrix`/`Mat` one would assume it is compatible to that as well.
Alternatively you could declare them as `explicit` constructors.

Reply:
> They both create a completely new Vector.

Ok, I can see that this is an argument why they should be in the Vector class.

The name instance is misleading though. Nouns usually denote properties and especially `instance` is a very common choice for the instance of a singleton.
These functions allocate a similar, or compatible Vector, so let's call them `allocate`.

It feels odd for the aforementioned reasons, but I'll put them in an overload set.